invokeai/backend/stable_diffusion/diffusion/custom_attention.py

```diff
@@ -0,0 +1,386 @@
+
+            if self._ip_adapter_attention_weights[ipa_index].skip:
+                continue
+
+            ipa_weights = self._ip_adapter_attention_weights[ipa_index].ip_adapter_weights
+            ipa_scale = regional_ip_data.scales[ipa_index]
+            ip_mask = ip_masks[0, ipa_index, ...]
+
+            # Expected ip_hidden_state shape: (batch_size, num_ip_images, ip_seq_len, ip_image_embedding)
+            ip_key = ipa_weights.to_k_ip(ip_hidden_states)
+            ip_value = ipa_weights.to_v_ip(ip_hidden_states)
+
+            ip_hidden_states = self.run_attention(
+                attn=attn,
+                query=query,
+                key=ip_key,
+                value=ip_value,
+                attention_mask=None,
+            )
+
+            # Expected ip_hidden_states shape: (batch_size, query_seq_len, num_heads * head_dim)
+            hidden_states = hidden_states + ipa_scale * ip_hidden_states * ip_mask
+
+        return hidden_states
+
+    def _get_slice_size(self, attn) -> Optional[int]:
+        if self.slice_size == "none":
+            return None
+        if isinstance(self.slice_size, int):
+            return self.slice_size
+
+        if self.slice_size == "max":
+            return 1
+        if self.slice_size == "balanced":
+            return max(1, attn.sliceable_head_dim // 2)
+
+        raise ValueError(f"Incorrect slice_size value: {self.slice_size}")
+
+    def run_attention(
+        self,
+        attn: Attention,
+        query: torch.Tensor,
+        key: torch.Tensor,
+        value: torch.Tensor,
+        attention_mask: Optional[torch.Tensor],
+    ) -> torch.Tensor:
+        slice_size = self._get_slice_size(attn)
+        if slice_size is not None:
+            return self.run_attention_sliced(
+                attn=attn,
+                query=query,
+                key=key,
+                value=value,
+                attention_mask=attention_mask,
+                slice_size=slice_size,
+            )
+
+        if self.attention_type == "torch-sdp":
+            attn_call = self.run_attention_sdp
+        elif self.attention_type == "normal":
+            attn_call = self.run_attention_normal
+        else:
+            raise Exception(f"Unknown attention type: {self.attention_type}")
+
+        return attn_call(
+            attn=attn,
+            query=query,
+            key=key,
+            value=value,
+            attention_mask=attention_mask,
+        )
+
+    def run_attention_normal(
+        self,
+        attn: Attention,
+        query: torch.Tensor,
+        key: torch.Tensor,
+        value: torch.Tensor,
+        attention_mask: Optional[torch.Tensor],
+    ) -> torch.Tensor:
+        query = attn.head_to_batch_dim(query)
+        key = attn.head_to_batch_dim(key)
+        value = attn.head_to_batch_dim(value)
+
+        attention_probs = attn.get_attention_scores(query, key, attention_mask)
```

Comment:
Raised https://github.com/huggingface/diffusers/pull/9075 to fix the type issue for `attention_mask` arg